File: python/atelier_bench/reports/technician_payroll.py
"""Technician payroll report."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..contracts.report_contract import ReportRow, row_from_cents
from ..support.abstract_report import AbstractReport

if TYPE_CHECKING:
    from ..dataset import Dataset


class TechnicianPayrollReport(AbstractReport):
    SLUG = "technician-payroll"

    def __init__(self) -> None:
        super().__init__(self.SLUG, "Technician payroll")

    def rows(self, data: "Dataset") -> list[ReportRow]:
        share = (data.labour_minutes() // max(len(data.technicians), 1)) * 125
        return [row_from_cents(technician.name, share) for technician in data.technicians]

File: go/atelier/reports/technician_payroll.go
// Package reports carries the 24 concrete reports. Each one embeds a base
// from atelier/support, declares only Rows, and registers itself in an init()
// function; Slug, Title, CacheKey and TTLSeconds arrive by promotion.
package reports

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction: this assertion is the ONLY textual
// link between the struct and the interface it satisfies.
var _ contracts.ReportContract = (*TechnicianPayrollReport)(nil)

// TechnicianPayrollReport reports technician payroll.
type TechnicianPayrollReport struct {
	support.BaseReport
}

// NewTechnicianPayrollReport builds the report with its slug and title.
func NewTechnicianPayrollReport() *TechnicianPayrollReport {
	return &TechnicianPayrollReport{BaseReport: support.NewBaseReport("technician-payroll", "Technician payroll")}
}

// Rows is the only contract method this type declares.
func (r *TechnicianPayrollReport) Rows(data *dataset.Dataset) []contracts.ReportRow {
	headcount := len(data.Technicians)
	if headcount == 0 {
		headcount = 1
	}
	share := int64(data.LabourMinutes()/headcount) * 125
	rows := make([]contracts.ReportRow, 0, len(data.Technicians))
	for _, technician := range data.Technicians {
		rows = append(rows, contracts.RowFromCents(technician.Name, share))
	}
	return rows
}

// init registers this report; the package-level registry is built by the 24
// init functions in this package, in file-name order.
func init() {
	Register(NewTechnicianPayrollReport())
}
